rust/lance-encoding/src/decoder.rs

```diff
@@ -0,0 +1,733 @@
+        // each column.  We take the column with the least amount of queued data and decrement that amount from the queued
+        // rows total of all columns.
+        while rows_to_read > 0 {
+            let mut min_rows_added = u32::MAX;
+            for (col_idx, field_scheduler) in self.field_schedulers.iter().enumerate() {
+                let status = &mut field_status[col_idx];
+                if status.rows_queued == 0 {
+                    let mut next_page = &field_scheduler[status.page_offset as usize];
+
+                    while status.rows_to_skip > next_page.num_rows() as u64 {
+                        status.rows_to_skip -= next_page.num_rows() as u64;
+                        status.page_offset += 1;
+                        next_page = &field_scheduler[status.page_offset as usize];
+                    }
+
+                    let page_range_start = status.rows_to_skip as u32;
+                    let page_rows_remaining = next_page.num_rows() - page_range_start;
+                    let rows_to_take = status.rows_to_take.min(page_rows_remaining as u64) as u32;
+                    let page_range = page_range_start..(page_range_start + rows_to_take);
+
+                    let scheduled = next_page.schedule_range(page_range, scheduler)?;
+
+                    status.rows_queued += rows_to_take as u64;
+                    status.rows_to_take -= rows_to_take as u64;
+                    status.page_offset += 1;
+                    status.rows_to_skip = 0;
+
+                    sink.send(scheduled).await.unwrap();
+
+                    min_rows_added = min_rows_added.min(rows_to_take);
+                }
+            }
+            if min_rows_added == 0 {
+                panic!("Error in scheduling logic, panic to avoid infinite loop");
+            }
+            rows_to_read -= min_rows_added as u64;
+            for field_status in &mut field_status {
+                field_status.rows_queued -= min_rows_added as u64;
+            }
+        }
+        Ok(())
+    }
+}
+
+// Represents the work to decode a single batch of data.  This is a CPU
+// driven task where we untangle any sophisticated encodings.
+struct DecodeBatchTask {
+    columns: Vec<Vec<Box<dyn DecodeArrayTask>>>,
+    schema: Schema,
+}
+
+impl DecodeBatchTask {
+    fn run(self) -> Result<RecordBatch> {
+        let columns = self
+            .columns
+            .into_iter()
+            .map(|col_tasks| {
+                let arrays = col_tasks
+                    .into_iter()
+                    .map(|col_task| col_task.decode())
+                    .collect::<Result<Vec<_>>>()?;
+                let array_refs = arrays.iter().map(|arr| arr.as_ref()).collect::<Vec<_>>();
+                // TODO: If this is a primtiive column we should be able to avoid this
+                // allocation + copy with "page bridging" which could save us a few CPU
+                // cycles.
```

Comment:
End-to-end zero copy would only be available in the case where we can't use any kind of encoding at all.  However, that may be the case for our FSL vectors since embeddings are pretty compressed already.

Otherwise, some kind of buffer copy is going to be inevitable because the shorter encoded representation is going to have to be decoded into a longer buffer.  Trying to do that in place is probably more work than its worth.

Either way, we can measure and see.  This is something we can iterate on if decoding starts becoming a CPU bottleneck.